csharp/src/Google.Protobuf/CodedInputStream.cs

```diff
@@ -711,7 +737,126 @@ public bool MaybeConsumeTag(uint tag)
+                {
+                    throw InvalidProtocolBufferException.InvalidWrapperMessageLength();
+                }
+                // field=1, type=64-bit = tag of 9
+                if (input.buffer[input.bufferPos + 1] != expectedTag)
+                {
+                    throw InvalidProtocolBufferException.InvalidWrapperMessageTag();
+                }
+                var result = BitConverter.ToDouble(input.buffer, input.bufferPos + 2);
+                input.bufferPos += 10;
+                return result;
+            }
+            else
+            {
+                int length = input.ReadLength();
+                if (length == 0)
+                {
+                    return 0D;
+                }
+                if (length != expectedLength)
+                {
+                    throw InvalidProtocolBufferException.InvalidWrapperMessageLength();
+                }
+                if (input.ReadTag() != expectedTag)
+                {
+                    throw InvalidProtocolBufferException.InvalidWrapperMessageTag();
+                }
+                return input.ReadDouble();
+            }
+        }
+
+        internal static double? ReadDoubleWrapperBigEndian(CodedInputStream input)
+        {
+            int length = input.ReadLength();
+            if (length == 0)
+            {
+                return 0D;
+            }
+            // tag:1 + value:8 = 9 bytes
+            if (length != 9)
+            {
+                throw InvalidProtocolBufferException.InvalidWrapperMessageLength();
+            }
+            // field=1, type=64-bit = tag of 9
+            if (input.ReadTag() != 9)
+            {
+                throw InvalidProtocolBufferException.InvalidWrapperMessageTag();
+            }
+            return input.ReadDouble();
+        }
+
+        internal static long? ReadInt64Wrapper(CodedInputStream input)
+        {
+            // field=1, type=varint = tag of 8
+            const int expectedTag = 8;
+            // length:1 + tag:1 + value:10(varint64-max) = 12 bytes
+            if (input.bufferPos + 12 <= input.bufferSize)
+            {
+                int length = input.buffer[input.bufferPos++];
+                if (length == 0)
+                {
+                    return 0L;
+                }
+                // Length will always fit in a single byte.
+                if (length >= 128)
+                {
+                    throw InvalidProtocolBufferException.InvalidWrapperMessageLength();
+                }
+                int finalBufferPos = input.bufferPos + length;
+                if (input.buffer[input.bufferPos++] != expectedTag)
+                {
+                    throw InvalidProtocolBufferException.InvalidWrapperMessageTag();
+                }
+                var result = input.ReadInt64();
+                // Verify this message only contained a single field.
+                if (input.bufferPos != finalBufferPos)

```

Comment:
A comment that the if statement on line 814 guarantees that ReadInt64 won't switch bufferpos.

Reply:
Done